torch/_higher_order_ops/out_dtype.py

```diff
@@ -0,0 +1,151 @@
+
+import torch
+import torch.utils._pytree as pytree
+from torch.fx.experimental.proxy_tensor import (
+    disable_proxy_modes_tracing,
+    ProxyTorchDispatchMode,
+    track_tensor_tree,
+)
+from torch.utils._python_dispatch import (
+    _get_current_dispatch_mode,
+    _pop_mode_temporarily,
+)
+from torch._C import DispatchKey
+from torch._functorch.eager_transforms import (
+    _unwrap_all_tensors_from_functional,
+    _wrap_all_tensors_to_functional,
+)
+from torch._ops import HigherOrderOperator
+from torch._subclasses.fake_tensor import FakeTensorMode
+from torch._prims_common import elementwise_dtypes, ELEMENTWISE_TYPE_PROMOTION_KIND
+
+
+out_dtype = HigherOrderOperator("out_dtype")
+out_dtype.fallthrough(DispatchKey.PythonDispatcher)
+out_dtype.fallthrough(DispatchKey.PythonTLSSnapshot)
+out_dtype.fallthrough(DispatchKey.ADInplaceOrView)
+out_dtype.fallthrough(DispatchKey.BackendSelect)
+out_dtype.fallthrough(DispatchKey.AutocastCPU)
+
+
+def trace_out_dtype(proxy_mode, func_overload, op, output_dtype, *args):
+    if not isinstance(op, torch._ops.OpOverload):
+        raise ValueError("out_dtype's first argument must be an OpOverload")
+
+    with disable_proxy_modes_tracing():
+        casted_args = pytree.tree_map_only(
+            torch.Tensor, lambda arg: arg.to(dtype=output_dtype), args
+        )
+        out = op(*casted_args)
+
+    node_args = (op, output_dtype, *args)
+    proxy_args = pytree.tree_map(proxy_mode.tracer.unwrap_proxy, node_args)
+    out_proxy = proxy_mode.tracer.create_proxy(
+        "call_function", func_overload, proxy_args, {}, name="out_dtype"
+    )
+    return track_tensor_tree(out, out_proxy, constant=None, tracer=proxy_mode.tracer)
+
+
+@out_dtype.py_impl(DispatchKey.CompositeExplicitAutograd)
+def out_dtype_dense(
+    op: torch._ops.OpOverload,
+    output_dtype: torch.dtype,
+    *args
+):
+    if not isinstance(op, torch._ops.OpOverload):
+        raise ValueError("out_dtype's first argument must be an OpOverload")
+
+    flat_inputs = pytree.tree_flatten(args)[0] + [torch.ones(1, dtype=output_dtype)]
+    promote_dtype: torch.dtype = elementwise_dtypes(
+        *flat_inputs,
+        type_promotion_kind=ELEMENTWISE_TYPE_PROMOTION_KIND.DEFAULT,
+    )[0]
+
+    casted_args = pytree.tree_map_only(
+        torch.Tensor, lambda arg: arg.to(dtype=promote_dtype), args
+    )
+    res = op(*casted_args).to(dtype=output_dtype)
+    return res
+
+
+@out_dtype.py_impl(DispatchKey.Autograd)
+def out_dtype_autograd(
+    op: torch._ops.OpOverload,
+    output_dtype: torch.dtype,
+    *args
+):
+    if not isinstance(op, torch._ops.OpOverload):
+        raise ValueError("out_dtype's first argument must be an OpOverload")
+
+    # TODO: support autograd
+    flat_operands, _ = pytree.tree_flatten(args)
+    assert all(
+        not f.requires_grad for f in flat_operands if isinstance(f, torch.Tensor)
+    ), "Autograd is not supported for out_dtype"
```

Comment:
Should be RuntimeError or NotImplementedError. Also, we should also check the grad_mode (if we are under torch.no_grad, then we don't need to compute autograd, so we do not need to error loudly here).